src/nbl/ext/DebugDraw/CDrawAABB.cpp

```diff
@@ -0,0 +1,374 @@
+
+	video::IGPUGraphicsPipeline::SCreationParams params[1] = {};
+	params[0].layout = layout;
+	params[0].vertexShader = vertex;
+	params[0].fragmentShader = fragment;
+	params[0].cached = {
+		.primitiveAssembly = {
+			.primitiveType = asset::E_PRIMITIVE_TOPOLOGY::EPT_LINE_LIST,
+		}
+	};
+	params[0].renderpass = renderpass;
+
+	device->createGraphicsPipelines(nullptr, params, &pipeline);
+
+	return pipeline;
+}
+
+bool DrawAABB::renderSingle(IGPUCommandBuffer* commandBuffer)
+{
+	commandBuffer->setLineWidth(1.f);
+	commandBuffer->draw(24, 1, 0, 0);
+
+	return true;
+}
+
+bool DrawAABB::render(IGPUCommandBuffer* commandBuffer, ISemaphore::SWaitInfo waitInfo, float* cameraMat3x4)
+{
+	using offset_t = SCachedCreationParameters::streaming_buffer_t::size_type;
+	constexpr auto MdiSizes = std::to_array<offset_t>({ sizeof(float32_t3), sizeof(InstanceData) });
+	// shared nPoT alignment needs to be divisible by all smaller ones to satisfy an allocation from all
+	constexpr offset_t MaxAlignment = std::reduce(MdiSizes.begin(), MdiSizes.end(), 1, [](const offset_t a, const offset_t b)->offset_t {return std::lcm(a, b); });
+	// allocator initialization needs us to round up to PoT
+	const auto MaxPOTAlignment = roundUpToPoT(MaxAlignment);
+
+	auto* streaming = m_cachedCreationParams.streamingBuffer.get();
+
+	auto* const streamingPtr = reinterpret_cast<uint8_t*>(streaming->getBufferPointer());
+	assert(streamingPtr);
+
+	commandBuffer->bindGraphicsPipeline(m_pipeline.get());	// move outside of loop, only bind once
+	commandBuffer->setLineWidth(1.f);
+
+	auto instancesIt = m_instances.begin();
+	const uint32_t verticesByteSize = sizeof(float32_t3) * m_unitAABBVertices.size();
+	const uint32_t availableInstancesByteSize = streaming->getBuffer()->getSize() - verticesByteSize;
+	const uint32_t instancesPerIter = availableInstancesByteSize / sizeof(InstanceData);
+	using suballocator_t = core::LinearAddressAllocatorST<offset_t>;
+	while (instancesIt != m_instances.end())
+    {
+		const uint32_t instanceCount = min(instancesPerIter, m_instances.size());
+        offset_t inputOffset = 0u;
+	    offset_t ImaginarySizeUpperBound = 0x1 << 30;
+	    suballocator_t imaginaryChunk(nullptr, inputOffset, 0, roundUpToPoT(MaxAlignment), ImaginarySizeUpperBound);
+	    uint32_t vertexByteOffset = imaginaryChunk.alloc_addr(verticesByteSize, sizeof(float32_t3));
+	    uint32_t instancesByteOffset = imaginaryChunk.alloc_addr(sizeof(InstanceData) * instanceCount, sizeof(InstanceData));
+	    const uint32_t totalSize = imaginaryChunk.get_allocated_size();
+
+	    inputOffset = SCachedCreationParameters::streaming_buffer_t::invalid_value;
+		std::chrono::steady_clock::time_point waitTill = std::chrono::steady_clock::now() + std::chrono::milliseconds(1u);
+	    streaming->multi_allocate(waitTill, 1, &inputOffset, &totalSize, &MaxAlignment);
+
+	    memcpy(streamingPtr + vertexByteOffset, m_unitAABBVertices.data(), sizeof(m_unitAABBVertices[0]) * m_unitAABBVertices.size());
```

Comment:
make a small buffer at startup containing the indices (please draw indexed) so you're not copying it in all the time

P.S. the vertices should be a `const float32_t3[8]` array in the vertex shader